securestorage2/src/main/java/de/adorsys/android/securestorage2/SecureStorage.kt

```diff
@@ -0,0 +1,269 @@
+    fun clearAllValuesAndDeleteKeys(context: Context) {
+        val applicationContext = context.applicationContext
+
+        checkAppCanUseLibrary()
+
+        if (KeyStoreTool.keyExists(applicationContext)) {
+            KeyStoreTool.deleteKey(applicationContext)
+        }
+        clearAllSecureValues(applicationContext)
+    }
+
+    @Throws(SecureStorageException::class)
+    fun registerOnSecureStorageChangeListener(
+        context: Context,
+        listener: SharedPreferences.OnSharedPreferenceChangeListener
+    ) {
+        val applicationContext = context.applicationContext
+
+        checkAppCanUseLibrary()
+
+        getSharedPreferencesInstance(applicationContext).registerOnSharedPreferenceChangeListener(listener)
+    }
+
+    @Throws(SecureStorageException::class)
+    fun unregisterOnSecureStorageChangeListener(
+        context: Context,
+        listener: SharedPreferences.OnSharedPreferenceChangeListener
+    ) {
+        val applicationContext = context.applicationContext
+
+        checkAppCanUseLibrary()
+
+        getSharedPreferencesInstance(applicationContext).unregisterOnSharedPreferenceChangeListener(listener)
+    }
+
+    internal fun getSharedPreferencesInstance(context: Context): SharedPreferences {
+        val applicationContext = context.applicationContext
+
+        return applicationContext.getSharedPreferences(
+            SecureStorageConfig.INSTANCE.SHARED_PREFERENCES_NAME,
+            MODE_PRIVATE
+        )
+    }
+
+    private fun putSecureValue(context: Context, key: String, value: String) {
+        getSharedPreferencesInstance(context).edit().putString(key, value)
+            .execute(SecureStorageConfig.INSTANCE.ASYNC_OPERATION)
+    }
+
+    private fun getSecureValue(context: Context, key: String): String? {
+        return getSharedPreferencesInstance(context).getString(key, null)
+    }
+
+    private fun removeSecureValue(context: Context, key: String) {
+        getSharedPreferencesInstance(context).edit().remove(key).execute(SecureStorageConfig.INSTANCE.ASYNC_OPERATION)
+    }
+
+    private fun clearAllSecureValues(context: Context) {
+        getSharedPreferencesInstance(context).edit().clear().execute(SecureStorageConfig.INSTANCE.ASYNC_OPERATION)
+    }
+
+    @Throws(SecureStorageException::class)
+    private fun checkAppCanUseLibrary() {
+        if (!SecureStorageConfig.INSTANCE.CAN_USE_LIBRARY) {
+            throw SecureStorageException(
+                "Cannot use SecureStorage2 on this device because it does not have hardware support.",
+                null,
+                SecureStorageException.ExceptionType.KEYSTORE_NOT_SUPPORTED_EXCEPTION
+            )
+        }
+    }
+}
+
+//================================================================================
+// SecureStorage Extension Function
+//================================================================================
+
+internal fun SharedPreferences.Editor.execute(async: Boolean) {
```

Comment:
In SecureStorageConfig the parameter ASYNC_OPERATION already is `true` by default. I moved that parameter to the method instead of always passing it as an argument.